### ISSUE TITLE
assigning a `Tuple[int, ...]` to an `Iterable[int]` results in `builtins.None` or `<nothing>`

### ISSUE THREAD

**Issue by carljm** [python/mypy#4466]
Bug with this code:

```
from typing import Iterable, Tuple
x: Iterable[int] = ()
y: Tuple[int, ...] = (1, 2, 3)
x = y
reveal_type(x)
```

Actual result (on mypy master, Python 3.6):

`tuplebug.py:9: error: Revealed type is 'builtins.None'`

Actual result with `--strict-optional`:

`tuplebug.py:9: error: Revealed type is '<nothing>'`

Expected result:

The revealed type should be `Tuple[int, ...]`.

Via git bisect, I find that this bug was introduced recently, in https://github.com/python/mypy/pull/4238 (cc @elazarg, @JukkaL ).

**Comment by carljm** [comment]
(We saw this as a regression false positive on real code, when upgrading mypy.)

**Comment by gvanrossum** [comment]
Ideally we should fix this before the next mypy release (which hasn't been scheduled yet).

**Comment by elazarg** [comment]
Thanks for the report.

I'd note that the following also fails, and is important to test since `Tuple[int, ...]` and `Tuple[int, int]` are handled quite differently inside mypy:
```python
from typing import Iterable, Tuple
x: Iterable[int] = ()
y: Tuple[int, int] = (1, 2)
x = y
reveal_type(x)  # should be Tuple[int, int], actual 'None'/'<nothing>'
```

The issue seems to be subtle.

**Comment by ethanhs** [comment]
@ilevkivskyi I think this is the same issue I saw in #5303. I am raising priority to high, since this was a regression and came up again (and I expect it is not an uncommon pattern).
